### ISSUE TITLE
rpi-eeprom-update doesn't not apply the latest firmware

### ISSUE THREAD

**Issue by goldyfruit** [raspberrypi/rpi-eeprom#462]
### Describe the bug

There is a new firmware available and as usual I wanted to apply this one but `rpi-eeprom-update` doesn't apply the new firmware, it says it did but when I reboot it's the old one.

When I run `sudo rpi-eeprom-update -a` in debug, I can see something wrong (full trace here):
```
sudo bash -x rpi-eeprom-update -a
+ set -e
+++ dirname rpi-eeprom-update
++ cd .
++ pwd
+ script_dir=/home/pi
+ '[' -f /etc/default/rpi-eeprom-update ']'
+ . /etc/default/rpi-eeprom-update
++ FIRMWARE_RELEASE_STATUS=beta
+ LOCAL_MODE=0
+ '[' -n '' ']'
+ '[' -d /lib/firmware/raspberrypi/bootloader ']'
+ FIRMWARE_ROOT=/lib/firmware/raspberrypi/bootloader
+ FIRMWARE_RELEASE_STATUS=beta
+ FIRMWARE_IMAGE_DIR=/lib/firmware/raspberrypi/bootloader/beta
+ FIRMWARE_BACKUP_DIR=/var/lib/raspberrypi/bootloader/backup
+ ENABLE_VL805_UPDATES=1
+ RECOVERY_BIN=/lib/firmware/raspberrypi/bootloader/beta/recovery.bin
+ BOOTFS=/boot
+ CM4_ENABLE_RPI_EEPROM_UPDATE=0
+ RPI_EEPROM_UPDATE_CONFIG_TOOL=raspi-config
+ RPI_EEPROM_SELF_UPDATE=0
+ RPI_EEPROM_SELF_UPDATE_MIN_VER=1650968668
+ BOOTLOADER_AUTO_UPDATE_MIN_VERSION=1599135103
+ DT_BOOTLOADER_TS=/proc/device-tree/chosen/bootloader/build-timestamp
+ EXIT_SUCCESS=0
+ EXIT_UPDATE_REQUIRED=1
+ EXIT_FAILED=2
+ EXIT_EEPROM_FROZEN=3
+ OVERWRITE_CONFIG=0
+ BOOTLOADER_FIRST_VERSION=1557513636
+ EEPROM_SIZE=524288
+ BOARD_INFO=
+ BOARD_REVISION=
+ BOARD_TYPE=
+ HAVE_VL805_EEPROM=0
+ TMP_EEPROM_IMAGE=
+ TMP_BOOTFS_MNT=
+ VL805_CURRENT_VERSION=
+ VL805_UPDATE_VERSION=
+ ACTION_UPDATE_BOOTLOADER=0
+ ACTION_UPDATE_VL805=0
+ CHECKSUMS=
+ trap cleanup EXIT
+ BOOTLOADER_CURRENT_VERSION=
+ BOOTLOADER_UPDATE_IMAGE=
+ BOOTLOADER_UPDATE_VERSION=0
+ AUTO_UPDATE_BOOTLOADER=0
+ AUTO_UPDATE_VL805=0
+ SILENT_UPDATE=0
+ MACHINE_OUTPUT=
+ JSON_OUTPUT=no
+ IGNORE_DPKG_CHECKSUMS=0
+ PACKAGE_INFO_DIR=/var/lib/dpkg/info/
+ '[' '!' -d /var/lib/dpkg/info/ ']'
+ getopts A:abdhilf:m:ju:rs option
+ case "${option}" in
+ AUTO_UPDATE_BOOTLOADER=1
+ AUTO_UPDATE_VL805=1
+ getopts A:abdhilf:m:ju:rs option
+ checkDependencies
+ '[' -f /sys/firmware/devicetree/base/system/linux,revision ']'
++ od -v -An -t x1 /sys/firmware/devicetree/base/system/linux,revision
++ tr -d ' \n'
+ BOARD_INFO=00d03114
+ '[' 1 -eq 0 ']'
+ '[' 3 -ne 3 ']'
+ BOARD_TYPE=17
+ BOARD_REVISION=4
+ '[' 17 -eq 20 ']'
+ '[' 17 -eq 17 ']'
+ '[' 4 -lt 4 ']'
+ HAVE_VL805_EEPROM=0
+ command -v rpi-eeprom-digest
+ command -v lspci
+ '[' -f /proc/device-tree/chosen/bootloader/build-timestamp ']'
+ '[' '!' -e /lib/firmware/raspberrypi/bootloader/beta ']'
+ '[' -e /lib/firmware/raspberrypi/bootloader/beta-00d03114 ']'
+ getBootloaderConfig
+ local blconfig_alias=/sys/firmware/devicetree/base/aliases/blconfig
+ local blconfig_nvmem_path=
+ '[' -f /sys/firmware/devicetree/base/aliases/blconfig ']'
++ strings /sys/firmware/devicetree/base/aliases/blconfig
+ local blconfig_ofnode_path=/sys/firmware/devicetree/base/reserved-memory/nvram@0
++ find -L /sys/bus/nvmem -samefile /sys/firmware/devicetree/base/reserved-memory/nvram@0
+ local blconfig_ofnode_link=/sys/bus/nvmem/devices/rmem0/of_node
+ '[' -e /sys/bus/nvmem/devices/rmem0/of_node ']'
++ dirname /sys/bus/nvmem/devices/rmem0/of_node
+ blconfig_nvmem_path=/sys/bus/nvmem/devices/rmem0
+ '[' -n /sys/bus/nvmem/devices/rmem0 ']'
+ cat /sys/bus/nvmem/devices/rmem0/nvmem
+ command -v sha256sum
+ '[' '!' -f /lib/firmware/raspberrypi/bootloader/beta/recovery.bin ']'
+ '[' 1 = 1 ']'
+ getBootloaderConfig
+ local blconfig_alias=/sys/firmware/devicetree/base/aliases/blconfig
+ local blconfig_nvmem_path=
+ grep FREEZE_VERSION=1
+ '[' -f /sys/firmware/devicetree/base/aliases/blconfig ']'
++ strings /sys/firmware/devicetree/base/aliases/blconfig
+ local blconfig_ofnode_path=/sys/firmware/devicetree/base/reserved-memory/nvram@0
++ find -L /sys/bus/nvmem -samefile /sys/firmware/devicetree/base/reserved-memory/nvram@0
+ local blconfig_ofnode_link=/sys/bus/nvmem/devices/rmem0/of_node
+ '[' -e /sys/bus/nvmem/devices/rmem0/of_node ']'
++ dirname /sys/bus/nvmem/devices/rmem0/of_node
+ blconfig_nvmem_path=/sys/bus/nvmem/devices/rmem0
+ '[' -n /sys/bus/nvmem/devices/rmem0 ']'
+ cat /sys/bus/nvmem/devices/rmem0/nvmem
+ checkAndApply
+ lookupVersionInfo
+ getBootloaderCurrentVersion
+ '[' -f /proc/device-tree/chosen/bootloader/build-timestamp ']'
+++ od /proc/device-tree/chosen/bootloader/build-timestamp -v -An -t x1
+++ tr -d ' '
++ printf %d 0x6390a20f
+ BOOTLOADER_CURRENT_VERSION=1670423055
+ '[' -z 1670423055 ']'
+ getBootloaderUpdateVersion
+ BOOTLOADER_UPDATE_VERSION=0
+ match='.*/pieeprom-[0-9][0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9].bin'
++ find /lib/firmware/raspberrypi/bootloader/beta/ -maxdepth 1 -type f -size 524288c -regex '.*/pieeprom-[0-9][0-9][0-9][0-9]-[0-9][0-9]-[0-9][0-9].bin'
++ head -n1
++ sort -r
+ latest=/lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin
+ '[' -f /lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin ']'
++ strings /lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin
++ grep BUILD_TIMESTAMP
++ sed 's/.*=//g'
+ BOOTLOADER_UPDATE_VERSION=1672828069
+ BOOTLOADER_UPDATE_IMAGE=/lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin
+ getVL805CurrentVersion
+ VL805_CURRENT_VERSION=
++ id -u
+ '[' 0 = 0 ']'
++ lspci -d 1106:3483 -xxx
++ awk '/^50:/ { print "VL805 FW version: " $5 $4 $3 $2 }'
+ vlver='VL805 FW version: 000138a1'
+ '[' -n 'VL805 FW version: 000138a1' ']'
+ VL805_CURRENT_VERSION=000138a1
+ ACTION_UPDATE_BOOTLOADER=0
+ ACTION_UPDATE_VL805=0
+ '[' 1672828069 -gt 1670423055 ']'
+ ACTION_UPDATE_BOOTLOADER=1
+ '[' 0 = 1 ']'
+ '[' 0 = 1 ']'
+ VL805_UPDATE_VERSION=000138a1
+ ACTION_UPDATE_VL805=0
+ removePreviousUpdates
++ id -u
+ '[' 0 = 0 ']'
+ findBootFS
+ blkid
+ grep -qE '/dev/mmcblk0p1.*LABEL_FATBOOT.*RECOVERY.*TYPE.*vfat'
+ '[' -d /boot ']'
++ find /boot/ -name '*.elf'
++ wc -l
+ '[' 8 = 0 ']'
+ rm -f /boot/recovery.bin
+ rm -f /boot/pieeprom.bin /boot/pieeprom.upd /boot/pieeprom.sig
+ rm -f /boot/vl805.bin /boot/vl805.sig
+ find /boot -maxdepth 1 -type f -follow -iname 'recovery.*' -regex '.*\.[0-9][0-9][0-9]$' -exec rm -f '{}' ';'
+ '[' 1 '!=' 1 ']'
+ '[' 1 '!=' 1 ']'
+ '[' 1 = 1 ']'
+ echo '*** INSTALLING EEPROM UPDATES ***'
*** INSTALLING EEPROM UPDATES ***
+ echo ''

+ printVersions
+ '[' 1 = 1 ']'
+ echo 'BOOTLOADER: update available'
BOOTLOADER: update available
++ date -u -d@1670423055
+ echo '   CURRENT: Wed 07 Dec 2022 02:24:15 PM UTC (1670423055)'
   CURRENT: Wed 07 Dec 2022 02:24:15 PM UTC (1670423055)
++ date -u -d@1672828069
+ echo '    LATEST: Wed 04 Jan 2023 10:27:49 AM UTC (1672828069)'
    LATEST: Wed 04 Jan 2023 10:27:49 AM UTC (1672828069)
+ echo '   RELEASE: beta (/lib/firmware/raspberrypi/bootloader/beta)'
   RELEASE: beta (/lib/firmware/raspberrypi/bootloader/beta)
+ echo '            Use raspi-config to change the release.'
            Use raspi-config to change the release.
+ echo ''

+ '[' 0 = 1 ']'
+ echo '  VL805_FW: Using bootloader EEPROM'
  VL805_FW: Using bootloader EEPROM
+ '[' 0 = 1 ']'
++ id -u
+ '[' 0 = 0 ']'
+ echo '     VL805: up to date'
     VL805: up to date
+ echo '   CURRENT: 000138a1'
   CURRENT: 000138a1
+ echo '    LATEST: 000138a1'
    LATEST: 000138a1
+ applyUpdate
++ id -u
+ '[' 0 = 0 ']'
+ '[' 0 = 0 ']'
+ package_checksums_file=/var/lib/dpkg/info//rpi-eeprom.md5sums
+ grep -qE '\.bin$' /var/lib/dpkg/info//rpi-eeprom.md5sums
++ mktemp
+ CHECKSUMS=/tmp/tmp.VyDNMxrORw
+ grep -E '\.bin$' /var/lib/dpkg/info//rpi-eeprom.md5sums
+ cd /
+ md5sum -c /tmp/tmp.VyDNMxrORw
+ applyRecoveryUpdate
+ '[' -n /lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin ']'
+ getBootloaderCurrentVersion
+ '[' -f /proc/device-tree/chosen/bootloader/build-timestamp ']'
+++ od /proc/device-tree/chosen/bootloader/build-timestamp -v -An -t x1
+++ tr -d ' '
++ printf %d 0x6390a20f
+ BOOTLOADER_CURRENT_VERSION=1670423055
+ '[' -z 1670423055 ']'
++ strings /lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin
++ grep BUILD_TIMESTAMP
++ sed 's/.*=//g'
+ BOOTLOADER_UPDATE_VERSION=1672828069
+ '[' 1670423055 -gt 1672828069 ']'
++ date -u -d@1670423055
+ echo '   CURRENT: Wed 07 Dec 2022 02:24:15 PM UTC (1670423055)'
   CURRENT: Wed 07 Dec 2022 02:24:15 PM UTC (1670423055)
++ date -u -d@1672828069
+ echo '    UPDATE: Wed 04 Jan 2023 10:27:49 AM UTC (1672828069)'
    UPDATE: Wed 04 Jan 2023 10:27:49 AM UTC (1672828069)
+ findBootFS
+ blkid
+ grep -qE '/dev/mmcblk0p1.*LABEL_FATBOOT.*RECOVERY.*TYPE.*vfat'
+ '[' -d /boot ']'
++ find /boot/ -name '*.elf'
++ wc -l
+ '[' 8 = 0 ']'
+ echo '    BOOTFS: /boot'
    BOOTFS: /boot
+ '[' -n /lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin ']'
+ '[' -f /lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin ']'
++ mktemp
+ TMP_EEPROM_IMAGE=/tmp/tmp.T4C4GBALlQ
+ prepareImage
+ '[' -f /lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin ']'
++ mktemp
+ TMP_EEPROM_IMAGE=/tmp/tmp.mBRSuU7naU
++ mktemp
+ TMP_EEPROM_CONFIG=/tmp/tmp.7nD88tmYRl
++ mktemp
+ NEW_EEPROM_CONFIG=/tmp/tmp.MXSyIs3I0u
+ mkdir -p /var/lib/raspberrypi/bootloader/backup
+ getBootloaderConfig
+ local blconfig_alias=/sys/firmware/devicetree/base/aliases/blconfig
+ local blconfig_nvmem_path=
+ '[' -f /sys/firmware/devicetree/base/aliases/blconfig ']'
++ strings /sys/firmware/devicetree/base/aliases/blconfig
+ local blconfig_ofnode_path=/sys/firmware/devicetree/base/reserved-memory/nvram@0
++ find -L /sys/bus/nvmem -samefile /sys/firmware/devicetree/base/reserved-memory/nvram@0
+ local blconfig_ofnode_link=/sys/bus/nvmem/devices/rmem0/of_node
+ '[' -e /sys/bus/nvmem/devices/rmem0/of_node ']'
++ dirname /sys/bus/nvmem/devices/rmem0/of_node
+ blconfig_nvmem_path=/sys/bus/nvmem/devices/rmem0
+ '[' -n /sys/bus/nvmem/devices/rmem0 ']'
+ cat /sys/bus/nvmem/devices/rmem0/nvmem
++ date +%Y%m%d-%H%M%S
+ backup=/var/lib/raspberrypi/bootloader/backup/pieeprom-backup-20230110-101622.conf
+ cp -f /tmp/tmp.7nD88tmYRl /var/lib/raspberrypi/bootloader/backup/pieeprom-backup-20230110-101622.conf
+ '[' -x '' ']'
+ cp -f /tmp/tmp.7nD88tmYRl /tmp/tmp.MXSyIs3I0u
++ wc -l /tmp/tmp.MXSyIs3I0u
++ awk '{print $1}'
+ '[' 6 -lt 3 ']'
+ cp -f /lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin /tmp/tmp.mBRSuU7naU
+ '[' 0 = 0 ']'
+ /home/pi/rpi-eeprom-config --out /tmp/tmp.mBRSuU7naU --config /tmp/tmp.MXSyIs3I0u /lib/firmware/raspberrypi/bootloader/beta/pieeprom-2023-01-04.bin
/usr/bin/rpi-eeprom-update: line 178: /home/pi/rpi-eeprom-config: No such file or directory
+ cleanup
+ '[' -f '' ']'
+ '[' -f /tmp/tmp.mBRSuU7naU ']'
+ rm -f /tmp/tmp.mBRSuU7naU
+ '[' -f /tmp/tmp.7nD88tmYRl ']'
+ rm -f /tmp/tmp.7nD88tmYRl
+ '[' -f /tmp/tmp.MXSyIs3I0u ']'
+ rm -f /tmp/tmp.MXSyIs3I0u
+ '[' -d '' ']'
+ TMP_BOOTFS_MNT=
+ TMP_EEPROM_IMAGE=
+ TMP_EEPROM_CONFIG=
+ NEW_EEPROM_CONFIG=
```

### Steps to reproduce the behaviour

`sudo rpi-eeprom-update -a`

### Device (s)

Raspberry Pi 4 Mod. B

### Bootloader configuration.

```
# rpi-eeprom-update 
*** UPDATE AVAILABLE ***
BOOTLOADER: update available
   CURRENT: Wed 07 Dec 2022 02:24:15 PM UTC (1670423055)
    LATEST: Wed 04 Jan 2023 10:27:49 AM UTC (1672828069)
   RELEASE: beta (/lib/firmware/raspberrypi/bootloader/beta)
            Use raspi-config to change the release.

  VL805_FW: Using bootloader EEPROM
     VL805: version unknown. Try sudo rpi-eeprom-update
   CURRENT: 
    LATEST: 
```

```
[all]
BOOT_UART=0
WAKE_ON_GPIO=1
POWER_OFF_ON_HALT=0
BOOT_ORDER=0xf14
DISABLE_HDMI=1
```

### System

```
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
```
vcgencmd bootloader_version
2022/12/07 14:24:15
version 0fc8fc8e5d5f506a83391f96e8285eb2a3116387 (release)
timestamp 1670423055
update-time 1673203016
capabilities 0x0000007f
```

```
Raspberry Pi reference 2021-05-07
Generated using pi-gen, https://github.com/RPi-Distro/pi-gen, 071d3956353f2be964d2d337d3b123f1872b7f53, stage2
```
```
Linux rpi4b01 5.15.84-v8+ #1613 SMP PREEMPT Thu Jan 5 12:03:08 GMT 2023 aarch64 GNU/Linux
```

```
# dpkg -l | grep -E "eeprom|bootloader"
ii  raspberrypi-bootloader               1:1.20230106-1                       arm64        Raspberry Pi bootloader
ii  raspberrypi-kernel                   1:1.20230106-1                       arm64        Raspberry Pi bootloader
ii  rpi-eeprom                           15.1-1                               arm64        Raspberry Pi 4 boot EEPROM updater
```

**Comment by timg236** [comment]
"/usr/bin/rpi-eeprom-update: line 178: /home/pi/rpi-eeprom-config: No such file or directory"

It looks like you have customized this or the paths somehow

**Comment by goldyfruit** [comment]
> "/usr/bin/rpi-eeprom-update: line 178: /home/pi/rpi-eeprom-config: No such file or directory"
> 
> It looks like you have customized this or the paths somehow

I didn't change anything, which is why I opened https://github.com/raspberrypi/rpi-eeprom/pull/461

**Comment by timg236** [comment]
"sudo bash -x rpi-eeprom-update -a"

The command should be "sudo rpi-eeprom-update -a"

**Comment by goldyfruit** [comment]
> "sudo bash -x rpi-eeprom-update -a"
> 
> The command should be "sudo rpi-eeprom-update -a"

Yep, you are correct, `bash` caused the issue.

**Comment by pelwell** [comment]
Running the script by invoking bash with its name as a parameter is doing something weird:
1. `sudo sh rpi-eeprom-update` does the correct(?) thing, saying `sh: 0: cannot open rpi-eeprom-update: No such file`, because `rpi-eeprom-update` is not a file in the current directory.
2. Running it the expected way, i.e. `sudo rpi-eeprom-update`, the shell searches the PATH for the script, patching $0 to be the path to that file.
3. Running it your way, i.e. `sudo bash rpi-eeprom-update`, bash must see that file doesn't exist in the current directory, and then search the PATH for it, but it doesn't set $0 as expected, which is why the script breaks.

In short, don't invoke this script (which happens to be a Bourne shell script, not a bash script) in that way. If you want the execution tracing, use:
```
$ sudo sh -x `which rpi-eeprom-update`
```
(or
```
$ sudo bash -x `which rpi-eeprom-update`
```
if you must use bash).

**Comment by timg236** [comment]
Thanks Phil!
